### PR TITLE
drivers/shtc1: Initial support for the SHTC1 temperature and humidity sensor

### PR DESCRIPTION
**Pull request by daexel**
### Contribution description
- SAUL integration added to this driver.
- Codefixes.

### Testing procedure
Testboard: SAMR21-xpro
Bildschirmfoto vom 2019-12-11 14-09-34
Bildschirmfoto vom 2019-12-11 14-10-33

### Issues/PRs references

Depends on PR #7866.

### CODE REVIEW

**Comment by benpicco** [comment]
#7866 hasn't seen activity in almost a year, so it's good you are taking this over :)

File: drivers/Makefile.dep [hunk 1]
```diff
@@ -663,5 +663,9 @@ endif
 ifneq (,$(filter dcf77,$(USEMODULE)))
   FEATURES_REQUIRED += periph_gpio
   FEATURES_REQUIRED += periph_gpio_irq
+endif
```
Comment (benpicco) [inline drivers/Makefile.dep]:
```suggestion
    USEMODULE += xtimer
endif
```

File: drivers/shtc1/include/shtc1_params.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#include "shtc1.h"
+#include "shtc1_regs.h"
+#include "saul_reg.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Default configuration parameters for SHTC1 sensors
```
Comment (aabadie) [inline drivers/shtc1/include/shtc1_params.h]:
```suggestion
 * @name    Default configuration parameters for SHTC1 sensors
```

File: drivers/shtc1/include/shtc1_params.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * Copyright 2017, RWTH Aachen. All rights reserved.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_shtc1
+ * @name        Default parameters for the SHTC1 Temperature and humidity sensor
```
Comment (aabadie) [inline drivers/shtc1/include/shtc1_params.h]: This line is not needed and should be removed.

File: tests/driver_shtc1/Makefile [hunk 4]
```diff
@@ -0,0 +1,9 @@
+include ../Makefile.tests_common
+
+DRIVER ?= shtc1
+
+USEMODULE += fmt
+USEMODULE += xtimer
```
Comment (aabadie) [inline tests/driver_shtc1/Makefile]: Not needed, xtimer is a dependency of the driver already.

File: drivers/shtc1/include/shtc1_regs.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+#ifndef SHTC1_REGS_H
+#define SHTC1_REGS_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* @name SHTC1 registers
```
Comment (aabadie) [inline drivers/shtc1/include/shtc1_regs.h]:
```suggestion
* @name SHTC1 registers
* @{
```

File: drivers/shtc1/include/shtc1_regs.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+*/
+#define SHTC1_CRC                                   (0x31)  /**< crc polynomial */
+#define SHTC1_MEASURE_CLOCK_STRETCHING_TEMP_HIGH    (0x7C)  /**< Clock stretching enable high*/
+#define SHTC1_MEASURE_CLOCK_STRETCHING_TEMP_LOW     (0xA2)  /**< Clock stretching disable high*/
+#define SHTC1_COMMAND_RESET_HIGH                    (0x80)  /**< Reset command high*/
+#define SHTC1_COMMAND_RESET_LOW                     (0x5D)  /**< Reset command low*/
+#define SHTC1_COMMAND_ID_HIGH                       (0xEF)  /**< Get ID command low*/
+#define SHTC1_COMMAND_ID_LOW                        (0xC8)  /**< Get ID command low*/
```
Comment (aabadie) [inline drivers/shtc1/include/shtc1_regs.h]:
```suggestion
#define SHTC1_COMMAND_ID_LOW                        (0xC8)  /**< Get ID command low*/
/** @} */
```

File: drivers/shtc1/include/shtc1_regs.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+* @name SHTC1 default ID
```
Comment (aabadie) [inline drivers/shtc1/include/shtc1_regs.h]:
```suggestion
* @brief SHTC1 default ID
```

File: drivers/include/shtc1.h [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+
+/**
+ * @brief   enumeration to enable or disable the crc-check
+ * @{
+ */
+typedef enum {
+    CRC_DISABLED = 0,   /**< Disable CRC Check */
+    CRC_ENABLED         /**< Enable CRC Check */
+} shtc1_crc_type_t;
```
Comment (aabadie) [inline drivers/include/shtc1.h]:
```suggestion
} shtc1_crc_type_t;
/** @} */
```

File: drivers/include/shtc1.h [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+ * @{
+ */
+typedef struct {
+    shtc1_values_t values;  /**< Values struct, where all read data will be stored. */
+    shtc1_params_t params;  /**< Parameters struct with all settings set. */
+} shtc1_t;
+/** @} */
+
+enum {
+    SHTC1_OK = 0,
```
Comment (aabadie) [inline drivers/include/shtc1.h]: Not documented

File: drivers/Makefile.include [hunk 10]
```diff
@@ -333,3 +333,7 @@ endif
 ifneq (,$(filter dcf77,$(USEMODULE)))
   USEMODULE_INCLUDES += $(RIOTBASE)/drivers/dcf77/include
 endif
+
+ifneq (,$(filter shtc1,$(USEMODULE)))
```
Comment (aabadie) [inline drivers/Makefile.include]: Please follow the alphabetical order of module includes

File: drivers/Makefile.dep [hunk 11]
```diff
@@ -663,5 +663,10 @@ endif
 ifneq (,$(filter dcf77,$(USEMODULE)))
   FEATURES_REQUIRED += periph_gpio
   FEATURES_REQUIRED += periph_gpio_irq
+    USEMODULE += xtimer
```
Comment (aabadie) [inline drivers/Makefile.dep]: unrelated change (2 extra spaces added)

File: drivers/include/shtc1.h [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+/**
+ * @brief   value struct, where all results obtained from the sensor will be saved.
+ * @{
+ */
+typedef struct {
+    int16_t temp;          /**< Temperature in centi °C (2372 = 23.72) after an call to the measurement function. */
+    uint16_t rel_humidity; /**< Relative humidity in centi percent (9922 = 99.22) after an call to the measuerment function. */
+    unsigned int id;       /**< ID read from the sensor, only available after shtc1_id() was called. */
+} shtc1_values_t;
```
Comment (benpicco) [inline drivers/include/shtc1.h]: Those are never used - and it doesn't really make sense to cache those values.
If you want to return both values without doing two reads over i2c, just add a 
```C
int8_t shtc1_read(const shtc1_t *dev, uint16_t *rel_humidity, uint16_t* temperature);
```
function.

Or maybe that's the only function and if you don't want either value you just pass `NULL`.

File: drivers/include/shtc1.h [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+/** @} */
+
+/**
+ * @brief   settings struct with all relevant parameters
+ * @{
+ */
+typedef struct {
+    i2c_t i2c_dev;              /**< I2C bus descriptor. */
+    uint8_t i2c_addr;           /**< I2C address of the sensor. */
+    shtc1_crc_type_t crc;       /**< crc check enabled or disabled (CRC_ENABLED/CRC_DISABLED). */
```
Comment (benpicco) [inline drivers/include/shtc1.h]: Does it make sense to have this run-time configurable?

File: drivers/include/shtc1.h [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+ * @brief Reads the ID and saves it in the device descriptor
+ *
+ * @details             When working correctly ID should equal xxxx'xxxx'xx00'0111 where x is unspecified.
+ *
+ * @param[in] dev       The I2C device descriptor.
+ *
+ * @return              SHTC1_OK on everything done.
+ * @return              SHTC1_ERROR on error.
+ */
+int8_t shtc1_id(shtc1_t* const dev);
```
Comment (benpicco) [inline drivers/include/shtc1.h]: How about

```suggestion
int8_t shtc1_id(shtc1_t* const dev, uint16_t *id);
```

File: drivers/include/shtc1.h (hunk 13 above)
Comment (daexel) [reply drivers/include/shtc1.h]: you mean the crc check?

Comment (benpicco) [reply drivers/include/shtc1.h]: Yes.
I'm just wondering why someone would set this on a per-device basis or be able to change this at run-time.

Making it a compile-time option should be enough imho.

Comment (daexel) [reply drivers/include/shtc1.h]: I would remove this because I think the crc-check belongs to a complete measurement...

File: drivers/shtc1/shtc1.c [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        crc ^= (rec_values[current_byte]);
+        for (bit = 8; bit > 0; bit--) {
+            if (crc & 0x80) {
+                crc = (crc << 1) ^ SHTC1_CRC;
+            }
+            else {
+                crc = (crc << 1);
+            }
+        }
+    }
```
Comment (benpicco) [inline drivers/shtc1/shtc1.c]: No need to implement this on your own.
If you rebase, you can just `#include "checksum/crc8.h"`

File: drivers/shtc1/shtc1.c [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    if (i2c_write_bytes(dev->params.i2c_dev, dev->params.i2c_addr, cmd, 2, 0)) {
+        return SHTC1_ERROR_BUS;
+    }
+    /* Receive the measurement */
+    /* 16 bit Temperature
+     * 8 bit  CRC temp
+     * 16 Bit Absolute Humidity
+     * 8 bit CRC Hum
+     */
+
```
Comment (benpicco) [inline drivers/shtc1/shtc1.c]: I think you want to remove that whitespace

File: drivers/shtc1/shtc1.c [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    /* Build and issue the measurement command */
+    uint8_t cmd[] =
+    { SHTC1_MEASURE_CLOCK_STRETCHING_TEMP_HIGH,
+      SHTC1_MEASURE_CLOCK_STRETCHING_TEMP_LOW };
+
+    i2c_acquire(dev->params.i2c_dev);
+    if (i2c_write_bytes(dev->params.i2c_dev, dev->params.i2c_addr, cmd, 2, 0)) {
+        return SHTC1_ERROR_BUS;
+    }
+    /* Receive the measurement */
```
Comment (benpicco) [inline drivers/shtc1/shtc1.c]: and move it here.

File: sys/auto_init/auto_init.c [hunk 18]
```diff
@@ -571,6 +575,7 @@ void auto_init(void)
     auto_init_veml6070();
 #endif
 
+
```
Comment (benpicco) [inline sys/auto_init/auto_init.c]: drop that

File: drivers/shtc1/shtc1.c [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+{
+    uint8_t received[6];
+    /* Build and issue the measurement command */
+    uint8_t cmd[] =
+    { SHTC1_MEASURE_CLOCK_STRETCHING_TEMP_HIGH,
+      SHTC1_MEASURE_CLOCK_STRETCHING_TEMP_LOW };
+
+    i2c_acquire(dev->params.i2c_dev);
+    if (i2c_write_bytes(dev->params.i2c_dev, dev->params.i2c_addr, cmd, 2, 0)) {
+
```
Comment (benpicco) [inline drivers/shtc1/shtc1.c]: I meant adding a new line after this if block and before the block comment below.
